From 654709230d8332ccf6656b4b451aec0c3f064c37 Mon Sep 17 00:00:00 2001
From: David Gobbi <david.gobbi@gmail.com>
Date: Tue, 17 Sep 2024 06:53:39 -0600
Subject: [PATCH] Add 'encoding' to vtkPythonInterpreter IO streams

When run in interactive mode under Python 3.13, the vtkpython exe
raised "AttributeError: 'vtkPythonStdStreamCaptureHelper' object has
no attribute 'encoding'" whenever it tried to print to the screen.

Adding an attribute "encoding='utf-8'" to the stream capture objects
fixes the exception.  Note that there is no check that the console is
actually UTF8, but VTK has historically assumed that it is.
---
 .../PythonInterpreter/vtkPythonStdStreamCaptureHelper.h    | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/Utilities/PythonInterpreter/vtkPythonStdStreamCaptureHelper.h b/Utilities/PythonInterpreter/vtkPythonStdStreamCaptureHelper.h
index 54c7476bf01..e162f359d05 100644
--- a/Utilities/PythonInterpreter/vtkPythonStdStreamCaptureHelper.h
+++ b/Utilities/PythonInterpreter/vtkPythonStdStreamCaptureHelper.h
@@ -16,7 +16,8 @@ VTK_ABI_NAMESPACE_BEGIN
 struct vtkPythonStdStreamCaptureHelper
 {
   PyObject_HEAD
-  int softspace; // Used by print to keep track of its state.
+  int softspace;        // Used by print to keep track of its state.
+  const char* Encoding; // Encoding, set to "utf-8"
   bool DumpToError;
 
   void Write(const char* string)
@@ -84,6 +85,9 @@ static PyMemberDef vtkPythonStdStreamCaptureHelperMembers[] = {
   { VTK_PYTHON_MEMBER_DEF_STR("softspace"), T_INT,
     offsetof(vtkPythonStdStreamCaptureHelper, softspace), 0,
     VTK_PYTHON_MEMBER_DEF_STR("Placeholder so print can keep state.") },
+  { VTK_PYTHON_MEMBER_DEF_STR("encoding"), T_STRING,
+    offsetof(vtkPythonStdStreamCaptureHelper, Encoding), READONLY,
+    VTK_PYTHON_MEMBER_DEF_STR("Text encoding for file.") },
   { nullptr, 0, 0, 0, nullptr }
 };
 
@@ -250,6 +254,7 @@ static vtkPythonStdStreamCaptureHelper* NewPythonStdStreamCaptureHelper(bool for
     PyObject_New(vtkPythonStdStreamCaptureHelper, &vtkPythonStdStreamCaptureHelperType);
   if (wrapper)
   {
+    wrapper->Encoding = "utf-8";
     wrapper->DumpToError = for_stderr;
   }
 
-- 
GitLab

